### PR TITLE
Gallery: Duplicate image IDs into easy-to-parse attribute

### LINKED ISSUE #8193: Core blocks Image and Gallery should provide easy access to their data
# Problem 

Text blocks, like paragraph and blockquote, can be styled just via CSS and would rarely need customizations of HTML markup.

Unlike text blocks, "Gallery" and "Image" might need advanced customization of HTML from the backend side.

## Use cases

- Replace `` with `<picture>`
- Replace `` with entirely different markup for the purpose of lazy-loading.
- Add special `sizes` attribute for responsive images.
- Add special attribute required by lightbox or another script (for example path to large image or dimensions of a large image)
- Specifically for the gallery - wrap a group of images with a specific element to create a specific layout (for example carousel or advanced grid).

Surely, all this would be applied by theme or by a plugin, so nothing should break when turning them on and off.

## Current Status of Image Block

Stores attachment ID in JSON.

```html
<!-- wp:image {"id":2253} -->
```

Which is good. But it does NOT store:

- Caption
- Textual Alternative option
- Image Size option (full, thumbnail, medium etc.)
- Link URL option (stores only linkDestination option)

## Current Status of Gallery Block

Gallery block does not store any data about images or their captions. The only way to retrieve it is to parse HTML via regex (specifically `data-id` attribute). Which is slow and unreliable.

# Solution Proposal

## 1. Store gallery attachment IDs in JSON

The Gallery should store an array with data about every image, specifically: attachment ID or image URL (if custom URL is added).

## 2. Do not separate native caption and alt from "Write caption" and "Textual Alternative" within Gutenberg

When editing Image caption in Gutenberg it should change the native WordPress Attachment "Caption" field too. Same for "Textual Alternative" (alt) option. They should also be named the same to avoid confusion.

image

### PR DESCRIPTION
**Pull request by mcsf**
## Description

Fixes #8193.
Would supersede #11212.
Prototyped during a call with @mtias.

**Warning:** yes, this is quite awful. I mainly want to close this unless I can be convinced this is a trade-off worth making.

This is yet another trade-off to allow easy parsing of galleries on the server. This functionality is expected by many, especially given the possibilities offered by the `[gallery]` shortcode, etc. A comprehensive set of server-side tools for deep block inspection isn't expected until WordPress 5.1, thus in the meantime measures such as this can make third-party developers' lives easier. See parent issue for more.

Conceptually, I don't like what this PR does, as it duplicates state. Indeed, Gallery has an attribute `images`, an array of image objects whose properties include `id`. This attribute is sourced in the block's HTML, however, based on elements' data attributes. This PR introduces an attribute `ids` which is not sourced and contains just the IDs extracted from `images`.

It highlights my concerns about state duplication:
- that state needs to be kept in sync with arguably fragile techniques, such as overriding `setAttributes` to enforce the unidirectional `images -> ids` synchronisation;
- that any piece that initialises a Gallery block needs to be aware of this duplication, as highlighted in the changes in `transforms` — but this would have to be replicated in *any other block* that offers a transform to a Gallery;
- this last point would not be properly solved by letting the Gallery constructor set its duplicate attribute automatically, as this would be yet another case of self-induced block changes that put the user in a state of ["stuck undo"](https://github.com/WordPress/gutenberg/issues/8119); it would also break synchronisation if a user somehow isn't in Visual/Block mode upon transforming blocks.

### CODE REVIEW

File: packages/block-library/src/gallery/index.js [hunk 1]
```diff
@@ -110,6 +115,12 @@ export const settings = {
 							} ) );
 						},
 					},
+					ids: {
+						type: 'array',
+						shortcode: ( { named: { ids } } ) => {
+							return ids.split( ',' );
```
Comment (aduth) [inline packages/block-library/src/gallery/index.js]: `ids` is not guaranteed to be assigned, at which point this would throw.

**Review by aduth — COMMENTED** [review]
I'm reasonably okay with this direction. If we have the IDs, I'd like to see `images` contain as little as possible which is not redundant with the `id`. I think that would mainly be captions? May even be possible to change attributes shape to `ids` and `captions` at that point. Though I suppose we'd probably want to limit such changes at this point in the development cycle.

The mention of a block "constructor" seems very similar to what we'd had early in Gutenberg with the function form of `attributes`, eventually removed due to its inability to interoperate with the backend (#1905).

**Comment by aduth** [comment]
>eventually removed due to its inability to interoperate with the backend (#1905).

Funny enough we still had at least one lingering reference in our documentation 😄 See #11765

File: packages/block-library/src/gallery/edit.js [hunk 2]
```diff
@@ -61,12 +61,28 @@ class GalleryEdit extends Component {
...
+		this.setAttributes = this.setAttributes.bind( this );
 
 		this.state = {
 			selectedImage: null,
 		};
 	}
 
+	setAttributes( attributes ) {
+		if ( attributes.ids ) {
+			throw new Error( "Don't use this attribute, dangit." );
```
Comment (mcsf) [inline packages/block-library/src/gallery/edit.js]: Note to self: even though this error should never be thrown, let's change the copy to something neutral and descriptive.

**Review by jorgefilipecosta — COMMENTED** [review]
I did some polishing on this PR; it should be ready. I think this approach is a good tradeoff between providing server access to the ids and using our existing mechanisms.
I agree with @aduth, that removing the id data attribute from the image may even be a better path. But I'm not sure if it is something we can do now, and this path seems "safer."

**Comment by antpb** [comment]
@jorgefilipecosta seems it is failing tests. Can't see why when comparing your changes to the last success...

```
FAIL test/e2e/specs/links.test.js (54.639s)
  ● Links › allows autocomplete suggestions to be selected with the mouse
    expect(received).not.toBeNull()
    Received: null
      302 | 		// Expect that clicking on the autocomplete suggestion doesn't dismiss the link popover.
      303 | 		await firstSuggestion.click();
    > 304 | 		expect( await page.$( '.editor-url-popover' ) ).not.toBeNull();
          | 		                                                    ^
      305 | 
      306 | 		// Expect the url input value to have been updated with the post url.
      307 | 		const inputValue = await page.evaluate( () => document.querySelector( '.editor-url-input input[aria-label="URL"]' ).value );
      at Object._callee17$ (test/e2e/specs/links.test.js:304:55)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:288:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at asyncGeneratorStep (test/e2e/specs/links.test.js:15:103)
      at _next (test/e2e/specs/links.test.js:17:194)
```

File: packages/block-library/src/gallery/index.js [hunk 3]
```diff
@@ -139,8 +156,12 @@ export const settings = {
 					mediaUpload( {
 						filesList: files,
 						onFileChange: ( images ) => {
+							const imagesAttr = images.map(
+								( image ) => pickRelevantMediaFiles( image )
```
Comment (mcsf) [inline packages/block-library/src/gallery/index.js]:
```suggestion
								pickRelevantMediaFiles
```

File: packages/block-library/src/gallery/index.js [hunk 4]
```diff
@@ -101,15 +116,17 @@ export const settings = {
...
-								id: parseInt( id, 10 ),
+							return parseShortCodeIds( ids ).map( ( id ) => ( {
+								id,
 							} ) );
 						},
 					},
+					ids: {
+						type: 'array',
+						shortcode: ( { named: { ids } } ) => {
+							return parseShortCodeIds( ids );
```
Comment (mcsf) [inline packages/block-library/src/gallery/index.js]:
```suggestion
							return parseShortcodeIds( ids );
```

File: packages/block-library/src/gallery/index.js [hunk 5]
```diff
@@ -101,15 +116,17 @@ export const settings = {
 					images: {
 						type: 'array',
 						shortcode: ( { named: { ids } } ) => {
-							if ( ! ids ) {
-								return [];
-							}
-
-							return ids.split( ',' ).map( ( id ) => ( {
-								id: parseInt( id, 10 ),
+							return parseShortCodeIds( ids ).map( ( id ) => ( {
```
Comment (mcsf) [inline packages/block-library/src/gallery/index.js]:
```suggestion
							return parseShortcodeIds( ids ).map( ( id ) => ( {
```

File: packages/block-library/src/gallery/index.js [hunk 6]
```diff
@@ -67,6 +71,16 @@ const blockAttributes = {
 
 export const name = 'core/gallery';
 
+const parseShortCodeIds = ( ids ) => {
```
Comment (mcsf) [inline packages/block-library/src/gallery/index.js]:
```suggestion
const parseShortcodeIds = ( ids ) => {
```

**Review by mcsf — COMMENTED** [review]
I suggested a couple of nit-picks, but LGTM :shipit:

**Comment by mcsf** [comment]
🚫 Actually, this still needs a deprecation for Gallery.

**Comment by jorgefilipecosta** [comment]
> 🚫 Actually, this still needs a deprecation for Gallery.

Hi @mcsf, in my tests I did not find any valid gallery created on the previous version that becomes invalid after this changes.
What I found was that we did not automatically add ids attribute to previous existing galleries, in order to improve that I used isEligible and added a migration.
This improvement also adds some nice features e.g. if we remove an image in the code (delete its li element) this mechanism also deletes the id attribute from the array, basically the mechanism makes sure images.id and ids are always in sync even if changed using the code editor.
I'm not sure if the reason you referred the need for deprecation was this one, if that is not the case feel free to let me know and I will update the logic.

**Review by youknowriad — APPROVED** [review]
The code looks good to me, and I wasn't able to trigger the deprecation error as well.

**Comment by mcsf** [comment]
> I'm not sure if the reason you referred the need for deprecation was this one, if that is not the case feel free to let me know and I will update the logic.

Actually, it was a quick phone-side comment and I didn't think that deeply, but this is great! Thanks for the work, @jorgefilipecosta!

**Comment by selftripsdotru** [comment]
Sorry, but problem is solved or not?

**Comment by noisysocks** [comment]
@selftripsdotru: Yes, this PR was merged and shipped as part of WordPress 5.0.

**Comment by selftripsdotru** [comment]
What is PR?
In the WordPress 5.0 or more, the problem is!!

**Comment by selftripsdotru** [comment]
How to get new IDs fo galleries after export/import?

**Comment by mcsf** [comment]
> What is PR?

A PR is a pull request, such as the one on this page. It means that this change has been incorporated in WordPress 5.0.

> How to get new IDs for galleries after export/import?

For general help requests, please post in the support forum at https://wordpress.org/support/forum/how-to-and-troubleshooting/.

Technical help requests have their own section of the support forum at https://wordpress.org/support/forum/wp-advanced/. 

You may also ask for technical support at https://wordpress.stackexchange.com/.

Please make sure you have checked the Handbook at https://wordpress.org/gutenberg/handbook before asking your question.

**Pull request merged**